File: src/scenes/Play.cpp
#include "Play.hpp"
#include "Pause.hpp"
#include "Over.hpp"
#include "CppInvaders.hpp"

PlayScene::PlayScene() {
    state = STARTING;
    horde_b = nullptr;
    ship_b = nullptr;

    for (int i = 0; i < 4; i++) {
        Shield s;
        s.deploy(32 + i * 48);
        shields.push_back(s);
    }
}

PlayScene::~PlayScene() {
    delete horde_b;
    delete ship_b;
}

void PlayScene::process_event(const Pico_Event &event) {
    switch (event.type) {
    case PICO_EVENT_KEY_DOWN:
        switch (event.key.keysym.sym) {
        case PICOK_SPACE:
            if (!ship_b && state == PLAYING && ship.state == Spaceship::DEPLOYED) {
                ship_b = ship.shoot();
            }
            break;
        case PICOK_ESCAPE:
            CppInvaders::get().scene = new PauseScene(this);
            break;
        }
        break;
    default:
        break;
    }
}

void PlayScene::process_collisions() {
    Pico_Anchor b_anc = PICO_ANCHOR_C;
    if (ship_b && ship_b->state == Bullet::ALIVE && horde_b && horde_b->state == Bullet::ALIVE) {
        Pico_Abs_Rect ship_b_rct = ship_b->get_rect();
        Pico_Abs_Rect horde_b_rct = horde_b->get_rect();
        if (SDL_HasIntersection(&ship_b_rct, &horde_b_rct)) {
            ship_b->explode(0.1);
            horde_b->explode(0.1);
        }
    }

    if (ship_b && ship_b->state == Bullet::ALIVE) {
        Pico_Abs_Rect ship_b_rct = ship_b->get_rect();

        if (ufo.collide_rect(ship_b_rct, b_anc)) {
            ufo.explode();
            ship_b->die(0.3);
        }

        int i = horde.collide_rect(ship_b_rct, b_anc);
        if (i >= 0) {
            int score = horde.kill_invader(i);
            ship_b->die(0.3);
            CppInvaders::get().add_to_score(score);
        }

        for (auto &s : shields) {
            if (s.damage(ship_b)) {
                ship_b->explode(0.1);
                break;
            }
        }
    }

    if (horde_b && horde_b->state == Bullet::ALIVE) {
        Pico_Abs_Rect horde_b_rct = horde_b->get_rect();
        if (ship.collide_rect(horde_b_rct, b_anc)) {
            ship.explode();
            horde_b->die(0.3);
        }
        
        for (auto &s : shields) {
            if (s.damage(horde_b)) {
                horde_b->explode(0.1);
                break;
            }
        }
    }

    for (int i = 0; i < 55; i++) {
        if (horde.invaders[i].is_alive()) {
            Pico_Abs_Rect inv_rct = horde.invaders[i].get_rect();
            for (auto &s : shields) {
                s.damage(inv_rct);
            }
        }
    }
}

void PlayScene::update(float delta) {
    switch (state) {
    case STARTING:
        horde.update(delta);
        ship.update(delta);
        if (horde.state == Horde::MARCHING) {
            state = PLAYING;
        }
        break;
    case PLAYING:
        if (horde_b) {
            horde_b->update(delta);
            if (horde_b->state == Bullet::DEAD) {
                delete horde_b;
                horde_b = nullptr;
            }
        } else if (horde.invaders_alive > 0 && ship.state == Spaceship::DEPLOYED) {
            horde_b = horde.shoot(ship.x);
        }

        if (ship_b) {
            ship_b->update(delta);
            if (ship_b->state == Bullet::DEAD) {
                delete ship_b;
                ship_b = nullptr;
            }
        }

        process_collisions();

        ufo.update(delta);
        horde.update(delta);
        ship.update(delta);

        if (horde.invaders_alive == 0 || ship.lives == 0) {
            state = FIN;
            timer = 0;
            delete horde_b;
            horde_b = nullptr;
            delete ship_b;
            ship_b = nullptr;
        }
        break;
    case FIN:
        timer += delta;
        if (timer >= 2 && ship.lives == 0) {
            CppInvaders::get().scene = new OverScene(this);
        } else if (timer >= 2) {
            state = STARTING;
            ufo = UFO();
            horde = Horde();
            ship = Spaceship();
            for (auto &s : shields) {
                s.deploy(s.x);
            }
        }
        break;
    }


}

void PlayScene::draw() const {
    ufo.draw();
    horde.draw();
    for (const auto &s : shields) {
        s.draw();
    }
    if (horde_b) {
        horde_b->draw();
    }
    if (ship_b) {
        ship_b->draw();
    }
    ship.draw();

    pico_set_color_draw(PICO_COLOR_WHITE);
    
    // Life counter at bottom-left
    char lives_buf[16];
    snprintf(lives_buf, sizeof(lives_buf), "%1d", ship.lives);
    Pico_Rel_Rect lives_txt_r = { '!', {8.0f, 248.0f, 0, 8}, PICO_ANCHOR_SW, NULL };
    pico_output_draw_text(lives_buf, &lives_txt_r);

    // Ship icons for remaining lives
    float ship_x = 24.0f;
    for (int i = 1; i < ship.lives; i++) {
        const char* path = IMG_SPACESHIP;
        Pico_Rel_Rect crop = { '!', {0, 0, 16, 8}, PICO_ANCHOR_NW, NULL };
        pico_layer_image_mode('=', "ship", path);
        pico_layer_sub_mode('=', "/crop/ship/life", "ship", &crop);
        Pico_Rel_Rect dst = { '!', {ship_x, 248.0f, 16, 8}, PICO_ANCHOR_SW, NULL };
        pico_output_draw_layer("/crop/ship/life", &dst);
        ship_x += 16.0f;
    }
}

File: scenes/Play.lua
local class = require "class"
local assets = require "assets"
local UFO = require "scenes.play.UFO"
local Horde = require "scenes.play.Horde"
local Spaceship = require "scenes.play.Spaceship"
local Shield = require "scenes.play.Shield"

local PlayScene = class.new()

function PlayScene:init(game_state)
    self.game_state = game_state
    self.state = "STARTING"
    self.horde_b = nil
    self.ship_b = nil

    self.shields = {}
    for i = 0, 3 do
        local s = Shield:new()
        s:deploy(32 + i * 48)
        table.insert(self.shields, s)
    end

    self.ufo = UFO:new()
    self.horde = Horde:new()
    self.ship = Spaceship:new(game_state)
    self.timer = 0
end

function PlayScene:process_event(event)
    if event.tag == 'key.dn' then
        if event.key == 'Space' then
            if not self.ship_b and self.state == "PLAYING" and self.ship.state == "DEPLOYED" then
                self.ship_b = self.ship:shoot()
            end
        elseif event.key == 'Escape' then
            local PauseScene = require "scenes.Pause"
            self.game_state.set_scene(PauseScene:new(self.game_state, self))
        end
    end
end

function PlayScene:process_collisions()
    local b_anc = 'C'
    if self.ship_b and self.ship_b.state == "ALIVE" and self.horde_b and self.horde_b.state == "ALIVE" then
        local sr = self.ship_b:get_rect()
        local hr = self.horde_b:get_rect()
        
        local x_min = math.max(sr.x, hr.x)
        local x_max = math.min(sr.x + sr.w, hr.x + hr.w)
        local y_min = math.max(sr.y, hr.y)
        local y_max = math.min(sr.y + sr.h, hr.y + hr.h)
        
        if x_min < x_max and y_min < y_max then
            self.ship_b:explode(0.1)
            self.horde_b:explode(0.1)
        end
    end

    if self.ship_b and self.ship_b.state == "ALIVE" then
        local sr = self.ship_b:get_rect()

        if self.ufo:collide_rect(sr, b_anc) then
            self.ufo:explode()
            self.ship_b:die(0.3)
        end

        local i = self.horde:collide_rect(sr, b_anc)
        if i >= 0 then
            local score = self.horde:kill_invader(i)
            self.ship_b:die(0.3)
            self.game_state.add_to_score(score)
        end

        for _, s in ipairs(self.shields) do
            if s:damage_shot(self.ship_b) then
                self.ship_b:explode(0.1)
                break
            end
        end
    end

    if self.horde_b and self.horde_b.state == "ALIVE" then
        local hr = self.horde_b:get_rect()
        if self.ship:collide_rect(hr, b_anc) then
            self.ship:explode()
            self.horde_b:die(0.3)
        end
        
        for _, s in ipairs(self.shields) do
            if s:damage_shot(self.horde_b) then
                self.horde_b:explode(0.1)
                break
            end
        end
    end

    for i = 1, 55 do
        local inv = self.horde.invaders[i]
        if inv and inv:is_alive() then
            local inv_rct = inv:get_rect()
            for _, s in ipairs(self.shields) do
                s:damage_rect(inv_rct)
            end
        end
    end
end

function PlayScene:update(delta)
    if self.state == "STARTING" then
        self.horde:update(delta)
        self.ship:update(delta)
        if self.horde.state == "MARCHING" then
            self.state = "PLAYING"
        end
    elseif self.state == "PLAYING" then
        if self.horde_b then
            self.horde_b:update(delta)
            if self.horde_b.state == "DEAD" then
                self.horde_b = nil
            end
        elseif self.horde.invaders_alive > 0 and self.ship.state == "DEPLOYED" then
            self.horde_b = self.horde:shoot(self.ship.x)
        end

        if self.ship_b then
            self.ship_b:update(delta)
            if self.ship_b.state == "DEAD" then
                self.ship_b = nil
            end
        end

        self:process_collisions()

        self.ufo:update(delta)
        self.horde:update(delta)
        self.ship:update(delta)

        if self.horde.invaders_alive == 0 or self.ship.lives == 0 then
            self.state = "FIN"
            self.timer = 0
            self.horde_b = nil
            self.ship_b = nil
        end
    elseif self.state == "FIN" then
        self.timer = self.timer + delta
        if self.timer >= 2 and self.ship.lives == 0 then
            local OverScene = require "scenes.Over"
            self.game_state.set_scene(OverScene:new(self.game_state, self))
        elseif self.timer >= 2 then
            self.state = "STARTING"
            self.ufo = UFO:new()
            self.horde = Horde:new()
            self.ship = Spaceship:new(self.game_state)
            for _, s in ipairs(self.shields) do
                s:deploy(s.x)
            end
        end
    end
end

function PlayScene:draw()
    self.ufo:draw()
    self.horde:draw()
    for _, s in ipairs(self.shields) do s:draw() end
    if self.horde_b then self.horde_b:draw() end
    if self.ship_b then self.ship_b:draw() end
    self.ship:draw()

    pico.set.color.draw("white")
    
    local lives_buf = tostring(self.ship.lives)
    local lives_txt_r = { '!', x=8.0, y=248.0, w=0, h=8, anchor='SW' }
    pico.output.draw.text(lives_buf, lives_txt_r)

    local ship_x = 24.0
    for i = 1, self.ship.lives - 1 do
        local crop = { '!', x=0, y=0, w=16, h=8, anchor='NW' }
        pico.layer.image('=', "ship", assets.IMG_SPACESHIP)
        pico.layer.sub('=', "/crop/ship/life", "ship", crop)
        local dst = { '!', x=ship_x, y=248.0, w=16, h=8, anchor='SW' }
        pico.output.draw.layer("/crop/ship/life", dst)
        ship_x = ship_x + 16.0
    end
end

return PlayScene
